### PR TITLE
Update local file and url download

### PR DESCRIPTION
**Pull request by Jodasue**
1. Updated the local file and made sure it was supported int TazerFile and lib.
2. Added support in cmake for CURL and SSL.  We must have SSL support to handle https files.  For now, we require both CURL and SSL.  If we find that we only care about http, then we could just use CURL.
3. Fixed some bugs in the python run script.  This script could still use some work, but is a strong start.

### CODE REVIEW

**Comment by Jodasue** [comment]
Added a second commit with a UrlDownload cache.  This cache downloads the whole file first, and then acts like a normal LocalFileCache.  This is a first approach... Next version we should not store the file in the file system, rather let the blocks live in the cache.

Broader comment, on if we want the downloading capability to be a cache?  Do we like the overloading of a regular cache method?  This means we have to change the code in the InputFile - to match LocalFile when adding caches.

File: src/client/Lib.cpp [hunk 1]
```diff
@@ -150,6 +151,7 @@ void __attribute__((constructor)) tazerInit(void) {
         timer.start();
         Loggable::mtx_cout = new std::mutex();
         InputFile::_cache = new Cache(BASECACHENAME, CacheType::base);
+        LocalFile::_cache = new Cache(BASECACHENAME, CacheType::base);
```
Comment (rdfriese) [inline src/client/Lib.cpp:158]: We might want to think about the implications of creating two separate caches...
do we want to allow both input and local at the same time?
thinking about it, can local just collapse down to inputfile just using "localfilecache/urlcache" as the llc?

File: src/client/LocalFile.cpp [hunk 2]
```diff
@@ -108,50 +111,55 @@
...
-    if (Config::LocalFileCache) {
-        c = BoundedFilelockCache::addNewLocalFileCache(LOCALFILECACHENAME);
-        std::cerr << "[TAZER] "
-                  << "filelock cache: " << (void *)c << std::endl;
+        c = SharedMemoryCache::addNewSharedMemoryCache(SHAREDMEMORYCACHENAME,CacheType::sharedMemory, Config::sharedMemoryCacheSize, Config::sharedMemoryCacheBlocksize, Config::sharedMemoryCacheAssociativity);
+        std::cerr << "[TAZER] " << "shared mem cache: " << (void *)c << std::endl;
         LocalFile::_cache->addCacheLevel(c, ++level);
     }
 
+    #ifdef USE_CURL
```
Comment (rdfriese) [inline src/client/LocalFile.cpp:134]: presumably we make this Config variable? (or this is set by CMAKE?

**Comment by rdfriese** [comment]
I agree with your first sentence, and think that we want to support both modes.

for the second sentence I think i might have asked something similar in one of the source comments I left. It might be reasonable to remove LocalFile altogether, and just fold that functionality into InputFile, and then just have the LocalFileCache or URLCache as the last level

**Comment by nathantallent** [comment]
As a general semantic comment, I do not think the download should be a cache: 1. Normal caches are bounded and are not guaranteed to hold an entire file. 2. Caches are normally used by multiple threads, so we would have to "pin" the blocks to ensure earlier parts of the file are not evicted while later parts are still downloading.

I could imagine reusing the implementation (if convenient) of an unbounded cache, which is perhaps what you were thinking.

> Added a second commit with a UrlDownload cache. This cache downloads the whole file first, and then acts like a normal LocalFileCache. This is a first approach... Next version we should not store the file in the file system, rather let the blocks live in the cache.
> 
> Broader comment, on if we want the downloading capability to be a cache? Do we like the overloading of a regular cache method? This means we have to change the code in the InputFile - to match LocalFile when adding caches.

**Pull request merged**